src/marker.rs

```diff
@@ -528,21 +286,123 @@ impl<'py> Python<'py> {
     /// }
     /// ```
     ///
+    /// # Example: The `send_wrapper` loophole is closed by running the closure on dedicated thread.
+    ///
+    /// ```should_panic
+    /// # use pyo3::prelude::*;
+    /// # use pyo3::types::PyString;
+    /// use send_wrapper::SendWrapper;
+    ///
+    /// Python::with_gil(|py| {
+    ///     let string = PyString::new(py, "foo");
+    ///
+    ///     let wrapped = SendWrapper::new(string);
+    ///
+    ///     py.allow_threads(|| {
+    ///         let sneaky: &PyString = *wrapped;
+    ///
+    ///         println!("{:?}", sneaky);
+    ///     });
+    /// });
+    /// ```
+    ///
     /// [`Py`]: crate::Py
     /// [`PyString`]: crate::types::PyString
     /// [auto-traits]: https://doc.rust-lang.org/nightly/unstable-book/language-features/auto-traits.html
     /// [Parallelism]: https://pyo3.rs/main/parallelism.html
     pub fn allow_threads<T, F>(self, f: F) -> T
     where
-        F: Ungil + FnOnce() -> T,
-        T: Ungil,
+        F: Send + FnOnce() -> T,
+        T: Send,
     {
+        use parking_lot::{const_mutex, Mutex};
+        use std::mem::{transmute, ManuallyDrop, MaybeUninit};
+        use std::panic::{catch_unwind, AssertUnwindSafe};
+        use std::sync::mpsc::{sync_channel, SendError, SyncSender};
+        use std::thread::{spawn, Result};
+        use std::time::Duration;
+
         // Use a guard pattern to handle reacquiring the GIL,
         // so that the GIL will be reacquired even if `f` panics.
         // The `Send` bound on the closure prevents the user from
         // transferring the `Python` token into the closure.
         let _guard = unsafe { SuspendGIL::new() };
-        f()
+
+        // To close soundness loopholes w.r.t. `send_wrapper` or `scoped-tls`,
+        // we run the closure on a separate thread so that it cannot
+        // access thread-local storage from the current thread.
+
+        // Construct a task
+        struct Task(*mut dyn FnMut());
+        unsafe impl Send for Task {}
+
+        let mut f = ManuallyDrop::new(f);
+        let mut result = MaybeUninit::<Result<T>>::uninit();
+
+        let (result_sender, result_receiver) = sync_channel(0);
+
+        let mut task = || {
+            // SAFETY: `F` is `Send` and we ensure that this closure is called at most once
+            let f = unsafe { ManuallyDrop::take(&mut f) };
+
+            result.write(catch_unwind(AssertUnwindSafe(f)));
+
+            result_sender.send(()).unwrap();
+        };
+        // SAFETY: the current thread will block until the closure has returned
+        let task = Task(unsafe { transmute(&mut task as &mut dyn FnMut()) });
+
+        // Enqueue task and spawn thread if necessary
+        static THREADS: Mutex<Vec<SyncSender<Task>>> = const_mutex(Vec::new());
```

Comment:
@davidhewitt @steffahn I added a second commit to re-use threads and as is to be expected by manual implementation of a scoped thread pool, there is quite a lot of unsafe involved about which I am somewhat unhappy. (I am especially nervous about unwinding as we need to keep the stack-bound data alive until the threads are done with it.)

Generally speaking, it does not seem nice to manually re-implement this but our requirements are somewhat special if I understand things correctly. For example, we never want to block waiting to re-use an existing thread, but rather spawn a new one if none are currently ready.

Finally, adding a tunable for the duration for which the threads wait for new tasks while idling is also not a welcome addition.

Reply:
I haven’t looked at the code yet, but one thing that came to mind is that the moment the helper thread is started, and the moment it ends, we are holding the GIL (which is a kind of mutex-like thing itself, if I understand correctly), so maybe that can serve to simplify synchronization concerns? Not that that would cut down on the amount of unsafe code…

For panics, one should probably best run `f()` with `catch_unwind`… then send the `Result` back to the thread where `allow_threads` was called, and [re-throw](https://doc.rust-lang.org/std/panic/fn.resume_unwind.html) the panic at the end, after the worker thread was put back into the queue.

Reply:
To fully and conservatively address concerns of accidentally leaving in place something in the main thread logic that can panic, during the section where the task was (or is about to be) sent, and the working thread not joined yet, it’s possible to insert a defensive abort bomb, i.e. a type that aborts the program (and informs the user that there’s a bug in `pyo3`) when dropped, and that gets defused after the worker thread’s response was received.

Reply:
> I hadn’t looked at the code yet before writing this first paragraph, but one thing that came to mind is that the moment the helper thread is started, and the moment it ends, we are holding the GIL (which is a kind of mutex-like thing itself, if I understand correctly), so maybe that can serve to simplify synchronization concerns? Not that that would cut down on the amount of unsafe code… so feel free to ignore this thought.

I considered this (i.e. `GILProtected<RefCell<..>>` instead of `Mutex<..>`) but decided against it as the was the possibility that the GIL would not yet be unlocked when the closure starts (on another thread) because it would be quite easy to accidentally still have it acquired when I send the task to a waiting thread. Hence, to keep things simple I chose to first release it and then do my own looking using a standard mutex.

> For panics, one should probably best run f() with catch_unwind… then send the Result back to the thread where allow_threads was called, and [re-throw](https://doc.rust-lang.org/std/panic/fn.resume_unwind.html) the panic at the end, after the worker thread was put back into the queue.

This is basically what I did, even though I did not use `resume_unwind` but rather just `unwrap`ed on a `std::thread::Result`. Will change that...

> To fully and conservatively address concerns of accidentally leaving in place something in the main thread logic that can panic, during the section where the task was (or is about to be) sent, and the working thread not joined yet, it’s possible to insert a defensive abort bomb, i.e. a type that aborts the program (and informs the user that there’s a bug in pyo3) when dropped, and that gets defused after the worker thread’s response was received.

Yes, I though about this and we can probably reuse our existing `PanicTrap` for that (which we currently use to avoid unwinding across the FFI boundary). Will add this...

Reply:
I don’t know how I overlooked the `catch_unwind`; probably because it’s so far away (in code) from the `spawn`.

Reply:
Seems like `ManuallyDrop` and `MaybeUninit` and the accompanying `unsafe` blocks are not actually necessary.

Reply:
It looks like we got to the point where the only unsafe code is the sharing of the function pointer to the stack data between the threads, nice! I suppose if we wanted to eliminate that we could use an allocation, but I personally like this way.

Reply:
> I suppose if we wanted to eliminate that we could use an allocation, but I personally like this way.

An extra allocation would remove the unsafe block when calling the tasks, but there would still be the unsafe block when transmuting the closure to lie to the compiler that it has a `+ 'static` bound (the trait object lifetimes are defaulted here and not immediately visible in the types) which cannot have because neither `F` nor `T` have a `+ 'static` bound.